Rigged to Blow: Lightbulb Rigging/Reagent Handling Overhaul

Rhials opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

First and foremost, guts out the previous, arguably cursed "rigged" functionality from light tubes. Light tubes now have a capacity of 20 reagents. Light tubes will now splash their contents upon shattering, instead of just releasing plasma gas into the air (if rigged). Lightbulbs with reagents inside of them will now transfer with them to the light fixture they're inserted into, which will **slowly heat the contents** to a maximum of 1000 degrees.

Heating the plasma reagent manually leads to it harmlessly boiling out of the container and dissipating, effectively ruining the original functionality. To account for plasma boiling over (when it should be bursting into flames) a SEALED_CONTAINER reagent container flag has been added. This is mostly just to prevent the plasma from leaving the container until it is ignition-ready. As a result, light tubes rigged the old way no longer harmlessly explode, and instead produce a small, (mostly) self-contained fire in the area surrounding the light.

Light tubes (and light fixtures with reagents in their lightbulb) now show their reagent contents upon examine. It's only fair to reward attentive people with a chance to avoid impending disaster, but on the other hand lightbulbs are transparent and you should just be able to look through them anyways.

As a bonus, all of this SHOULD be handled in a way that doesn't put any unnecessary strain on machine processing or anything. Rejoice.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

Lightbulbs being reagent holders was a huge missed opportunity, with their only use being for the disappointingly ineffective plasma-rigging functionality. This expands on the idea, and reworks the original functionality to be much more interactive. Bigger sandbox! Deeper sandbox! 

<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
add: Rigging light fixtures now works with more reagents than just plasma. Light fixtures will heat the reagents of their inserted lightbulb, up to a maximum of 1000 degrees. Lightbulb tubes now hold 20 reagents to make this more usable.
add: Lightbulbs will now splash their contents on whatever they're shattered by. Their contents are also now visible upon examination.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Rhials committed:
removes the rigging functionality

Rhials committed:
cleans up rigging from light replacer

Rhials committed:
planning, comments

Rhials committed:
why is that there

Rhials committed:
removes stray signal unregisters and changes a define

Rhials committed:
reagent holder for lights works now

Rhials committed:
heating works, number of lights running process() should be reduced

still trying to figure out the heat equation, and why the fuck delta_time doesn't go past 2

Rhials committed:
drop th e s

Rhials committed:
adds sealed_container framework

Rhials committed:
wowee it works now

this results in a brief, SELF CONTAINED plasmafire that reaches about 5 tiles in diameter. Pretty solid

Rhials committed:
adds transparency to lights/light holders, adds light replacer emag,

splash functionality works too now. These things are just glorified beakers now

Rhials committed:
I LOVE SPELLCHECK PRS

Rhials committed:
whoops

Rhials committed:
clean up code

Rhials committed:
number define

Rhials committed:
when did that get there

Zonespace27 commented on code/game/objects/items/devices/lightreplacer.dm:
```diff
@@ -203,14 +203,13 @@
-			target.rigged = (obj_flags & EMAGGED ? 1 : 0)
 			target.brightness = L2.brightness
+			if(obj_flags & EMAGGED)
+				target.create_reagents(20, SEALED_CONTAINER | TRANSPARENT) //LIGHT_REAGENT_CAPACITY can't be accessed from here. Look for a not magic-number solution to this please
```
the soultion would probably be moving the _light_defines.dm to the __DEFINES folder

Rhials committed:
moves light_defines to _DEFINES

Thank you to Zone "Zonespace27" Space27 for the help

Rhials committed:
get outta here

Rhials committed:
minor documentation changes

Kapu1178 commented on code/modules/power/lighting/light.dm:
```diff
@@ -313,17 +313,18 @@
+		if(length(light_object.reagents.reagent_list))
+			create_reagents(LIGHT_REAGENT_CAPACITY, SEALED_CONTAINER | TRANSPARENT)
+			light_object.reagents.copy_to(reagents, LIGHT_REAGENT_CAPACITY)
+			light_object.reagents.Destroy()
```
NO NO NO BAD BAD BAD BAD DO NOT DO THIS. NEVER manually call `Destroy()` 

`QDEL_NULL(light_object.reagents)`.

Kapu1178 commented on code/modules/power/lighting/light.dm:
```diff
@@ -542,8 +543,10 @@
 	var/obj/item/light/light_object = new light_type()
+	if(reagents)
+		reagents.copy_to(light_object.reagents, LIGHT_REAGENT_CAPACITY)
+		reagents.Destroy()
```
SCREAM

Rhials committed:
oh my goodness gracious

Wallemations commented:
How does the icon look for the reagent being visible inside the bulb?

Rhials commented:
> How does the icon look for the reagent being visible inside the bulb?

PR description wasn't clear, and has been edited for clarification. Nothing is visible on the sprite itself, but the light tubes/fixtures now show a "it contains X of various reagents" message like beakers and similar reagent holders.

Wallemations commented:
Ahhh I see

moocowswag commented:
The light bulb can have some gunpowder as a treat.

JohnFulpWillard commented on code/modules/power/lighting/light.dm:
```diff
@@ -241,13 +238,16 @@
 		return PROCESS_KILL
 	if(has_power())
-		if (cell.charge == cell.maxcharge)
+		if (cell.charge == cell.maxcharge && !reagents)
```
Why are we checking for !reagents twice?

Rhials replied:
Originally the only circumstances in which process() would fire was to affect the cell charge, or to handle low power mode. I tacked on the first check to prevent PROCESS_KILL from firing early if the emergency power cell wasn't present, but the light was still running from station power. Let me find a good way to rearrange this.

JohnFulpWillard commented on code/modules/power/lighting/light.dm:
```diff
@@ -542,8 +543,10 @@
 /obj/machinery/light/proc/drop_light_tube(mob/user)
 	var/obj/item/light/light_object = new light_type()
+	if(reagents)
+		reagents.copy_to(light_object.reagents, LIGHT_REAGENT_CAPACITY)
+		QDEL_NULL(reagents)
```
why not use ``trans_to()`` isntead?

Rhials replied:
The QDEL_NULL (at least on this line) ensures that the tradeoff between the light tube and the machine doesn't leave anything behind. Removing it causes every light with a reagent lightbulb to look like this after it gets removed, swapped, etc. 

![sleight of hand](https://user-images.githubusercontent.com/28870487/190525795-b76122e3-3f74-4dad-b611-c4f3a2f06850.PNG)

Rhials committed:
transfer two

Rhials committed:
I dont need this either I dont think

Rhials committed:
rewrites process() to be a bit more flexible. I think this should work?

Removed comments from line 246-247

//If it's still processing at this point, it's probably running on emergency power.
//The only two conditions that call SSmachines to start in update() are (if: reagents) or (if: has emergency power & is on)

JohnFulpWillard commented on code/modules/power/lighting/light_items.dm:
```diff
@@ -91,49 +89,35 @@
 /obj/item/light/create_reagents(max_vol, flags)
 	. = ..()
-	RegisterSignal(reagents, list(COMSIG_REAGENTS_NEW_REAGENT, COMSIG_REAGENTS_ADD_REAGENT, COMSIG_REAGENTS_DEL_REAGENT, COMSIG_REAGENTS_REM_REAGENT), .proc/on_reagent_change)
 	RegisterSignal(reagents, COMSIG_PARENT_QDELETING, .proc/on_reagents_del)
```
Do we need this signal at all? All it does is unregister itself now.

Rhials committed:
goodbye extraneous code

this was essentially just used for handling the signals related to rigging functionality. I didn't keep it in because I thought it might be important later. It wasn't.

private-tristan commented:
are you able to inject directly into the lightbulb? or do you need to take it out, inject reagents, and put it back in?

this seems really fun to use as a chemist, fill supercooled phlog and 1u stab. agent in lightbulbs and watch as the station erupts into flames a few minutes later

Rhials commented:
> are you able to inject directly into the lightbulb? or do you need to take it out, inject reagents, and put it back in?

You inject directly into the lightbulb, and shouldn't be able to inject anything into a light frame itself if a light is installed. The intent is to provide the same sleight-of-hand illusion as the original rigging functionality -- reagents are "stored" in the "bulb", even though the bulb is destroyed/remade on insert/removal. The only time a light frame should have a reagent holder attached is when it is inherited from the inserted light tube, which disappears and "moves" back to the light tube when you remove it.

JohnFulpWillard commented on code/modules/power/lighting/light.dm:
```diff
@@ -176,6 +174,8 @@
 		if(color)
 			color_set = color
+		if(reagents)
+			START_PROCESSING(SSmachines, src)
```
Small question I had in my dreams, why do we set it to process when we get reagents? Don't we only want it to get hot while the light switch is on, in which case it would process itself anyways?

Rhials replied:
From my understanding, process was only used to charge a depleted emergency cell. The light isn't actually running process() when it is on unless there's a cell that's being charged inside (or, in this PR, is heating reagents). Regardless, removing these lines stops the lights from heating under normal circumstances.
